### ISSUE TITLE
DCEVM for Java 17

### ISSUE THREAD

**Issue by Odex64** [dcevm/dcevm#198]
I was wondering if there will be a version for Java 17(LTS) since it is available only for Java 7, 8 and 11.

**Comment by skybber** [comment]
There will be.

**Comment by MrDandas** [comment]
@skybber do you have ETA ?

**Comment by coolsa** [comment]
Not sure if it counts as it isn't directly from this project, but I noticed that the jetbrains runtime beta versions for jdk17 seem to have dcevm options available, and it seems to work for my use case (Minecraft 1.17 fabric modding).

https://github.com/JetBrains/JetBrainsRuntime/releases/tag/jbr17b135.1
has options with download links `jbrsdk_dcevm-17-PLATFORM-bVERSION.tar.gz`, which when using with the VM argument `-XX:+AllowEnhancedClassRedefinition` enables alterations to the body of a class.
I haven't run any extensive testing on it to see any limitations or differences from earlier versions of DCEVM, but it seems to work wonderfully.


Double check for a more recent release of the jbr17 betas if you follow those instructions, but otherwise hopefully this helps anyone who needs to use dcevm for java 17.

**Comment by MrDandas** [comment]
I need to change class hierarchy and add/remove/modify interfaces. Will check the JetBrains. Thanks

**Comment by skybber** [comment]
Class hiearchy is not supported even in Jetbrains version. It has not been supported since dcevm7 and I did not port it to Jetbrains version.

**Comment by skybber** [comment]
Otherwise, interface addition/removal is supported.

**Comment by tfactor2** [comment]
@skybber Cool, thanks!
Do you know the future of Trava OpenJDK?
Should jbr be used instead?

**Comment by skybber** [comment]
Yes, jbr17 is recommended. Currently there is a bug in jbr-dcevm17, that will be fixed in the next release. Also there is an issue with HA, that is not embedded with jbr-runtime. From perspective of application developer using Spring/J2EE etc, HotswapAgent is "must have" feature.

**Comment by tfactor2** [comment]
So, _jbr with dcevm_ effectively cannot be used in the development process - the runtime doesn't support modifications of class signatures? Tried to add a method in an opened debug session both on jbr11 and jbr17 with dcevms with no luck:

> Hot Swap failed. dcevm intellij - ...: add method not implemented dcevm intellij ...: Operation not supported by V

**Comment by skybber** [comment]
You have to use either `-XX:+AllowEnhancedClassRedefinition` or `-XX:HotswapAgent=..` to enable advanced hotswap.

**Comment by tfactor2** [comment]
Thanks, works perfectly in both 11 and 17 versions!

Regarding http://hotswapagent.org/ - for me, the solution in general even on supported JDKs (Trava) didn't work. Tried on Hibernate; the implementation relied on Hibernate internals, just stopped working with newer Hibernate releases. Thus, we can't rely on such an unstable solution in a regular development process.
For us, enhanced class redefinition would be more than enough.

**Comment by skybber** [comment]
What version of Hibernate are you using? Please create issue at https://github.com/HotswapProjects/HotswapAgent,  Otherwise we have tests up to v5.4.14 https://github.com/HotswapProjects/HotswapAgent/blob/master/plugin/hotswap-agent-hibernate-plugin/run-tests.sh

**Comment by tfactor2** [comment]
@skybber, thanks for the quick response. 
Unfortunately, used it quite a while ago, so can't tell the exact version.
We plan to migrate to Java17, thus we anyways will have to stick to JBR w/o HotswapAgent support.

**Comment by Jaryt** [comment]
> Not sure if it counts as it isn't directly from this project, but I noticed that the jetbrains runtime beta versions for jdk17 seem to have dcevm options available, and it seems to work for my use case (Minecraft 1.17 fabric modding).
> 
> https://github.com/JetBrains/JetBrainsRuntime/releases/tag/jbr17b135.1 has options with download links `jbrsdk_dcevm-17-PLATFORM-bVERSION.tar.gz`, which when using with the VM argument `-XX:+AllowEnhancedClassRedefinition` enables alterations to the body of a class. I haven't run any extensive testing on it to see any limitations or differences from earlier versions of DCEVM, but it seems to work wonderfully.
> 
> Double check for a more recent release of the jbr17 betas if you follow those instructions, but otherwise hopefully this helps anyone who needs to use dcevm for java 17.

Minecraft Forge modder here. Unfortunately it doesn't seem that the bootstrap launcher used by Minecraft forge is enjoying the switch to Jbr17. I believe we need these features due to our mods using Mixins and it conflicting with the ability to even hotswap inner method changes.

**Comment by skybber** [comment]
Minecraft does not work with jbr17?

**Comment by Jaryt** [comment]
> Minecraft does not work with jbr17?

Turns out it was due to the Minecraft forge version, which has recently been updated!

**Comment by 1cg** [comment]
@skybber you are doing God's work keeping the DCEVM going.  It's insane that this isn't baked into the core JVM.

Do you think there will be an "official" jdk 17 release any time soon?

**Comment by skybber** [comment]
Dcevm17 is now part of jbr runtime, that is fork of JVM with a lot of enhancements. JB moved to Europe recently, I recommend to look at theirs official statement to current situation.

**Comment by 1cg** [comment]
Interesting.  Is that good news or bad news?

When I look here:

https://github.com/JetBrains/JetBrainsRuntime

I see jdk 11 releases.  Can you send me a link to their statement?

Regardless, thank you for keeping this incredibly important project alive!

**Comment by skybber** [comment]
Look at https://github.com/JetBrains/JetBrainsRuntime/releases , dcevm17 is embedded in JBR17, there isn't separate dcevm17 build. Enhanced hotswap must be explicitly turned on using `-XX:+AllowEnhancedClassRedefinition`

**Comment by t-oster** [comment]
If I understand it correctly you can add HotSwapAgent manually to the JBR, so reloading will also work with spring boot and hibernate etc (https://github.com/TravaOpenJDK/trava-jdk-11-dcevm/issues/46) or is this not needed?

**Comment by skybber** [comment]
yes, just put HA  into `lib/hotswap/`, and use `-XX:HotswapAgent=fatjar`

**Comment by 1cg** [comment]
That is *fantastic* I have it working locally.  Are you still associated with the project?

**Comment by skybber** [comment]
Yes, I'm involved in the project, JB is sponsoring dcevm now.

**Comment by TheBestPessimist** [comment]
do you know when the dcevm website will be updated with this information?

**Comment by skybber** [comment]
We have to add it to hotswapagent pages, in the same way how the page http://hotswapagent.org/mydoc_quickstart-jdk11.html  is done. Pages sources are at https://github.com/HotswapProjects/HotswapAgentPages, any help will be appreciated.

**Comment by TheBestPessimist** [comment]
And i think the readme here should also be updated to point out the jetbrains replacement.

**Comment by skybber** [comment]
Anyway, there are many dcevm fixes in JBR 11/17 now, but it's still not 100% stable, especially the rare case where adding/removing fields is difficult to solve if stack frames contain old instances replaced by scalars.  Maybe Oracle is aware of this and that's why they don't include dcevm in the standard hotspot, maybe the reason is completely different - they are working on a solution in Graal, but according to the specs this solution is far from better than dcevm.

**Comment by 1cg** [comment]
@skybber do you have a github sponsor or similar set up?  I'd like to sponsor your DCEVM work.

(Jetbrains recently sponsored me for htmx, and this is far more imporant work than that!)

**Comment by skybber** [comment]
Personally I don't have it, there quite a lot of people involved in HA+dcevm, it would be ... selfish to have sponsors just for myself.

**Comment by 1cg** [comment]
:) Is there a place where you all meet to discuss?  If so, and if you don't mind, it would be great to get this work sponsored.  

This is an extremely important feature for JVM languages and I'm sure many people in the community would be willing to chip in if there were a single organization to do so through!

**Comment by skybber** [comment]
pls. conntact me by an email.

**Comment by tfactor2** [comment]
Probably, worth noticing that the DCEVM is now integrated into the **standard** Jetbrains Runtime build according to https://github.com/JetBrains/JetBrainsRuntime/issues/134.

Thank you!

**Comment by skybber** [comment]
It is even used as a default debug enviroment for IntelliJ Idea development in Jetbrains now, it is also able to use it to run profiling tools like YourKit or JFR.

**Comment by jsalts** [comment]
Confirmed that this works with the jbr. I got tripped up because hotswap-agent.jar is not bundled in as far as I could tell like it used to be. You'll need the hotswap agent plugin for your IDE and then add "-XX:+AllowEnhancedClassRedefinition -XX:HotswapAgent=fatjar" to your VM options. You seem to be able to ignore messages about the plugin not being able to determine the dcevm version.

**Comment by gdrte** [comment]
Some one from Jetbrains advised -XX:HostswapAgent is not required any more as this is baked into the JBR itself.
https://youtrack.jetbrains.com/issue/JBR-5195/I-sometime-have-my-project-that-is-blocked-from-starting-and-i-have-the-feeling-that-there-is-a-deadlock-somwhere-during-the
@skybber Can you please clarify. 
Thank you!

**Comment by skybber** [comment]
HotswapAgent is not in JBR, if you need it you can either:

- put `hotswap-aget.jar` it into `$JAVA_HOME/lib/hotswap/` and use `-XX:HotswapAgent=fatjar`
- or use `-XX:HotswapAgent=external  -javaagent:<path>/hotswap-agent.jar`
